src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+                case SyntaxKind.InterfaceDeclaration:
+                    return ((InterfaceDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.EnumDeclaration:
+                    return ((EnumDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.DelegateDeclaration:
+                    return ((DelegateDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.EventFieldDeclaration:
+                    return ((EventFieldDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.MethodDeclaration:
+                    return ((MethodDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.OperatorDeclaration:
+                    return ((OperatorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.ConversionOperatorDeclaration:
+                    return ((ConversionOperatorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.ConstructorDeclaration:
+                    return ((ConstructorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.BaseConstructorInitializer:
+                case SyntaxKind.ThisConstructorInitializer:
+                    return ((ConstructorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.DestructorDeclaration:
+                    return ((DestructorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.PropertyDeclaration:
+                    return ((PropertyDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.IndexerDeclaration:
+                    return ((IndexerDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                case SyntaxKind.AddAccessorDeclaration:
+                    return ((AccessorDeclarationSyntax)currentNode).SemicolonToken.IsMissing;
+                default:
+                    // At this point, the node should be empty or its children should not end with a semicolon.
+                    Debug.Assert(!currentNode.ChildNodesAndTokens().Any()
+                        || !currentNode.ChildNodesAndTokens().Last().IsKind(SyntaxKind.SemicolonToken));
+                    return false;
+            }
+        }
+
+        /// <summary>
+        /// Determines if a statement ends with a closing delimiter, and that closing delimiter exists.
+        /// </summary>
+        /// <remarks>
+        /// <para>Statements such as <c>do { } while (expression);</c> contain embedded enclosing delimiters immediately
+        /// preceding the semicolon. These delimiters are not part of the expression, but they behave like an argument
+        /// list for the purposes of identifying relevant places for statement completion:</para>
```

Comment:
i don't really understand what this means.

Reply:
I can't think of a better wording here... is there a specific part that is not clear?